catalyst/contrib/nn/criterion/cluster.py

```diff
@@ -0,0 +1,164 @@
+        Returns:
+            modified matrix of shape (p, p - 1) created from matrix by
+            deletion of diagonal elements
+        """
+        p, _ = matrix.shape
+        indices = torch.diag(torch.ones(p)).bool()
+        diagonal_free = matrix[~indices].view((p, p - 1))
+        return diagonal_free
+
+    def _batch_hard_cluster_loss(
+        self, embeddings: torch.Tensor, labels: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Count cluster loss with hard sampling over the batch.
+        Each batch should contains k samples for p classes.
+
+        Args:
+            embeddings: tensor of shape (batch_size; embed_dim)
+            where batch_size = k * p
+            labels: labels of the batch, of size (batch_size,)
+
+        Returns:
+            scalar tensor containing the cluster loss
+        """
+        # Get matrix of indices of labels in batch
+        labels_mask = self._get_label_mask(labels)
+        p = labels_mask.shape[0]
+        k = labels.shape[0] // p
+
+        # Validate batch: expected to get batch
+        # with k samples for p classes
+        assert (labels_mask.sum(1, keepdim=True) == k).all(), ValueError(
+            f"For batch of shape {labels_mask.shape} with "
+            f"{p} classes required {k} samples for ech class."
+        )
+
+        embed_dim = embeddings.shape[-1]
+        # Reshape embeddings to groups of (p, k, embed_dim) ones,
+        # each i-th group contains embeddings of i-th class.
+        embeddings = embeddings.repeat((p, 1, 1))
+        embeddings = embeddings[labels_mask].view((p, -1, embed_dim))
+
+        # Count mean vectors for each class in batch
+        mean_vectors = embeddings.mean(1)
+
+        d_intra = self._count_intra_class_distances(embeddings, mean_vectors)
+        # Count the distances to the sample farthest from mean vector
+        # for each class.
+        d_intra = d_intra.max(1).values
+        # Count matrix of distances from mean vectors to each other
+        d_inter = self._count_inter_class_distances(mean_vectors)
+        # For each class mean vector get the closest mean vector
+        d_inter = self._skip_diagonal(d_inter)
+        d_inter = d_inter.min(1).values
+
+        # Count batch loss. For each class i:
+        # loss_i = max(d_intra - d_inter + margin, 0)
+        # loss = loss_1 + loss_2 + ... + loss_p
+        loss = torch.mean(torch.relu(d_intra - d_inter + self.margin))
+        return loss
+
+    def forward(
+        self, embeddings: torch.Tensor, targets: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Forward propagation method for cluster loss.
+
+        Args:
+            embeddings (torch.Tensor): tensor of shape (batch_size, embed_dim)
+            targets: labels of the batch, shape (batch_size,)
+
+        Returns:
+            torch.Tensor: cluster loss for the batch
+        """
+        return self._batch_hard_cluster_loss(embeddings, targets)
+
+
+__all__ = ["ClusterLoss"]
```

Comment:
could you please also add it to `catalyst/contrib/nn/criterion/__init__.py`?

Reply:
and to our docs here, https://github.com/catalyst-team/catalyst/blob/master/docs/api/contrib.rst#criterion